Season 03/Jungmin/Week 16/week16_4803.java

```diff
@@ -0,0 +1,69 @@
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.StringTokenizer;
+
+public class week16_4803 {
+    public static List<Integer>[] graph;
+    public static int[] visited;
+    public static int n, m;
+    public static boolean isCycle;
+    public static void main(String[] args) throws IOException {
+        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+        StringTokenizer st;
+        StringBuilder sb = new StringBuilder();
+
+        int caseNum = 0;
+        while(true) {
+            st = new StringTokenizer(br.readLine());
+            n = Integer.parseInt(st.nextToken());
+            m = Integer.parseInt(st.nextToken());
+
+            if (n == 0 && m == 0) break;
+            caseNum++;
+
+            graph = new ArrayList[n+1];
+            visited = new int[n+1];
+
+            for (int i = 0; i <= n; i++) graph[i] = new ArrayList<>();
+
+            for (int i = 0; i < m; i++) {
+                st = new StringTokenizer(br.readLine());
+                int a = Integer.parseInt(st.nextToken());
+                int b = Integer.parseInt(st.nextToken());
+                graph[a].add(b);
+                graph[b].add(a);
+            }
+
+            int tree = 0;
+            for (int i = 1; i <= n; i++) {
+                if (visited[i] == 0) {
+                    isCycle = false;
+                    dfs(i, 0);
+                    if (!isCycle) tree++;   // 사이클이 존재하지 않을 때에만 트리 개수를 카운트
+                }
+            }
+
+            if (tree == 0) sb.append("Case " + caseNum + ": No trees.").append('\n');
+            else if (tree == 1) sb.append("Case " + caseNum + ": There is one tree.").append('\n');
+            else sb.append("Case " + caseNum + ": A forest of " + tree + " trees.").append('\n');
+        }
+
+        System.out.println(sb);
+    }
+
+    public static void dfs(int i, int prev) {
+        visited[i] = 1;
+
+        for (int next : graph[i]) {
+            if (next == prev) continue; // 이전 노드로 되돌아가지 않음
+            if (visited[next] == 1) {   // 방문했던 노드라면 사이클이 존재하는 것
+                isCycle = true;
+                return;
+            }
+            if (visited[next] == 0) dfs(next, i);
+        }
+    }
```

Comment:
isCycle을 리턴하는 함수로 만들었다면 어땠을까요?! 사실근데 실전에 들어서면 그냥 전역변수가 사실맘편하긴해요

```suggestion
    public static boolean dfs(int i, int prev) {
        visited[i] = 1;
        
        for (int next : graph[i]) {
            if (next == prev) continue; // 이전 노드로 되돌아가지 않음
            if (visited[next] == 1) {   // 방문했던 노드라면 사이클이 존재하는 것
                return true; // isCycle = true;
            }
            if (visited[next] == 0) 
               if( dfs(next, i))  return true;
        }
        return false;
    }
```

Reply:
오 제 코드랑 로직이 같아서 변수명만 매칭해 읽었네요! 새삼 자바 기억이 흐릿함을 느끼고 갑니다....

Reply:
@yujin1292 isCycle로 다 구현해놓고 불 값을 리턴하는 함수 생각이 나더라구요 귀찮아서 그냥 안 바꾼건 비밀 하핫
@jeongiin 통했네여 역싀 역싀~!